communication/src/actor/message.rs

```diff
@@ -0,0 +1,212 @@
+    /// match the one that was expected or is otherwise invalid.
+    InvalidPeerId,
+    /// The connection was dropped because the connection limit
+    /// for a peer has been reached.
+    ConnectionLimit(ConnectionLimit),
+    /// An I/O error occurred on the connection.
+    IO,
+    /// The connection handler produced an error.
+    Handler,
+    /// Timout on connection attempt
+    Timeout,
+}
+
+impl<TTransErr> From<PendingConnectionError<TTransErr>> for ConnectPeerError {
+    fn from(error: PendingConnectionError<TTransErr>) -> Self {
+        match error {
+            PendingConnectionError::Transport(_) => ConnectPeerError::Transport,
+            PendingConnectionError::InvalidPeerId => ConnectPeerError::InvalidPeerId,
+            PendingConnectionError::ConnectionLimit(limit) => ConnectPeerError::ConnectionLimit(limit),
+            PendingConnectionError::IO(_) => ConnectPeerError::IO,
+        }
+    }
+}
+
+impl From<DialError> for ConnectPeerError {
+    fn from(error: DialError) -> Self {
+        match error {
+            DialError::Banned => ConnectPeerError::Banned,
+            DialError::ConnectionLimit(limit) => ConnectPeerError::ConnectionLimit(limit),
+            DialError::NoAddresses => ConnectPeerError::NoAddresses,
+        }
+    }
+}
+
+impl<THandlerErr> From<ConnectionError<THandlerErr>> for ConnectPeerError {
+    fn from(error: ConnectionError<THandlerErr>) -> Self {
+        match error {
+            ConnectionError::Handler(_) => ConnectPeerError::Handler,
+            ConnectionError::IO(_) => ConnectPeerError::IO,
+        }
+    }
+}
+
+/// Requests for the [`CommuncationActor`]
+#[derive(Debug, Clone)]
+pub enum CommunicationRequest<Req, T: Message> {
+    /// Send a request to a remote peer.
+    /// This requires that a connection to the targeted peer has been established and is active.
+    RequestMsg { peer_id: PeerId, request: Req },
+    /// Set the actor reference that incoming request are forwarded to.
+    SetClientRef(ActorRef<T>),
+    /// Connect to a remote peer.
+    /// If the peer id is know it will attempt to use a know address of it, otherwise the `addr` will be dialed.
+    ConnectPeer { addr: Multiaddr, peer_id: PeerId },
+    /// Check if a connection to that peer is currently active.
+    CheckConnection(PeerId),
+    /// Obtain information about the swarm.
+    GetSwarmInfo,
+    /// Ban a peer, which prevent any connection to that peer.
+    BanPeer(PeerId),
+    /// Unban a peer to allow future communication.
+    UnbanPeer(PeerId),
+    /// Start listening to a port on the swarm. If no `Multiaddr` is provided, the address will be OS assigned.ActorRef
+    StartListening(Option<Multiaddr>),
+    /// Stop listening to the swarm. Without a listener, the local peer can not be dialed from remote.
+    RemoveListener,
+}
+
+/// The firewall that rejected or dropped the request
+#[derive(Debug, Clone)]
+pub enum FirewallRejected {
```

Comment:
`FirewallRejected` sounds like an error type, what about `RejectionFirewall` or something like that? or just `FirewallType`

Reply:
How about `RejectionFirewallSource` or just `FirewallSource`? 
`FirewallType` sounds like there are different types of firewalls, but actually they differ on their source/ origin (i.g. is it the local firewall or the remote one), rather than on their type. But then again maybe "type" can be interpreted in different ways.